Add SpectroscopicFibreOptic and FibreOpticGroup observers

vsnever opened the pull request:
This adds `SpectroscopicFibreOptic` and `FibreOpticGroup` observers and changes the implementation of `SpectroscopicSightLine` and `LineOfSightGroup` to fix #283.

The following changes are made:

- `SpectroscopicSightLine` and `SpectroscopicFibreOptic` are now `SightLine` and `FibreOptic` instances, respectively;
- `SpectroscopicSightLine` and `SpectroscopicFibreOptic` can have multiple pipelines;
- Pipelines of the following classes  are supported: `PowerPipeline0D`, `RadiancePipeline0D`, `SpectralPowerPipeline0D`, `SpectralRadiancePipeline0D`;
- The property `observed_spectrum()` is converted to a method;
- `plot_spectra()` and `observed_spectrum()` take pipeline index or name as an argument;
- `LineOfSightGroup` and `FibreOpticGroup` are instances of the parent `Observer0DGroup` class;
- The properties of individual sight lines in a group are controlled through respective properties of the group;
- It is possible now to share the same property value across all sight lines in a group;
- A new method `connect_pipelines()`, which allows to connect the pipelines with the same class, name and filter (optionally) to all sight lines in the group, is added; 
- A few bugs are fixed.

Note that until https://github.com/raysect/source/pull/386 and https://github.com/raysect/source/pull/388 are merged, the `display_progress` property is not working and the mono pipelines cannot be accessed by name.

Mateasek commented:
I'm sorry I missed this issue. I have one idea about the FibreOpticGroup. I hope I understood correctly the point it offering an easy manipulation with a set of observers of the same type, right? What about allowing the observers to have different properties at the end? For example some observation volumes can be approximated with FibreOptics but with slightly different radii and acceptance angles. They could however benefit from all the other methods the FibreOpticGroup offers.

vsnever commented:
> I'm sorry I missed this issue.

This issue was added just a couple of hours ago, so you haven't missed it :).

> What about allowing the observers to have different properties at the end? 

The observers within the group can have different properties. It is possible to change the property of a certain observer by accessing this observer directly:
```python
# by name (if unique)
group['sight line 2'].acceptance_angle = 5
# by index
group[1].acceptance_angle = 5
# or via sight_line list
group.sight_lines[1].acceptance_angle = 5
```
or by using group properties:
```python
# assuming the group has 3 sight lines this will set different acceptance angles for all 3 of them
group.acceptance_angle = [5., 7., 9.]
```
However, this:
```python
group.acceptance_angle = 5.
```
will set the acceptance angle to 5 deg for all sight lines in the group.

Mateasek commented:
Perfect! This will be really useful!

vsnever commented:
Just noticed a couple of safety issues that were also present in the original implementation. 

1. The `_sight_lines` attribute must be a tuple, not a list. Otherwise, the sight lines can be added bypassing the checks.
2. Accessing the sight line by name should throw an error if the name is not unique.

vsnever commented:
>     1. The `_sight_lines` attribute must be a tuple, not a list. Otherwise, the sight lines can be added bypassing the checks.
> 
>     2. Accessing the sight line by name should throw an error if the name is not unique.

Fixed in last commit.

jacklovell commented on cherab/tools/observers/spectroscopic.py:
```diff
@@ -440,28 +444,29 @@ def sight_lines(self):
 
         sight_line.parent = self
-        self._sight_lines.append(sight_line)
+        new_sight_lines = list(self._sight_lines)
+        new_sight_lines.append(sight_line)
+        self._sight_lines = tuple(new_sight_lines)
```
There is a cleaner way to do this:
```python
self._sight_lines = self._sight_lines + (sight_line,)
```
Tuples have an `__add__` method which will return a new tuple with the operand appended on the end.

vsnever replied:
Thanks. I fixed this.

Mateasek reviewed (commented):
Hi Vlad,

thanks for this PR, the groups will be definitely useful. I have a 2 questions which are rather about the concept than code.

The first is the limit to use only a single pipeline. I think it takes away the beauty of pipelines. There are many diagnostic systems and applications where you want to get the whole spectrum, but also some filtered and integrated parts at once (for example Thomson scattering, diagnostics based on spectral filters and fast detectors as diodes and photo-multipliers). It would probably make the handling of the classes a bit more difficult for users, but it would in my opinion make it much more useful and would be worth it.

vsnever commented:
Hi @Mateasek,

>The first is the limit to use only a single pipeline. I think it takes away the beauty of pipelines. There are many diagnostic systems and applications where you want to get the whole spectrum, but also some filtered and integrated parts at once (for example Thomson scattering, diagnostics based on spectral filters and fast detectors as diodes and photo-multipliers). It would probably make the handling of the classes a bit more difficult for users, but it would in my opinion make it much more useful and would be worth it.

>The second is if you decide only for a single pipeline, why does it have to be SpectralRadiancePipeline0D?

I totally agree with you that these observers would benefit from supporting multiple pipelines, but this is how they were originally designed by @mattngc. I have deliberately not made any changes that would require updating the existing code that uses these observers. If we allow these observers to support multiple pipelines, `observed_spectrum` can no longer be a property. We have to convert it to a method that gets the index or name of the pipeline and returns the corresponding spectrum. However, we can easily update the code we know about, like `load_ks5_sightlines()` in `cherab.jet.spectroscopy`. So I'll let them support multiple pipelines of type `Pipeline0D`, we just need to remember to include this in the changelog.

vsnever commented on cherab/tools/observers/spectroscopic.py:
```diff
@@ -101,105 +428,222 @@ def plot_spectra(self, unit='J', ymax=None):
     @property
     def point(self):
+        """
+        The observation point of the sight line.
```
I agree, let's rename it to `origin`. It was named that before me.

vsnever commented on cherab/tools/observers/spectroscopic.py:
```diff
@@ -101,105 +428,222 @@ def plot_spectra(self, unit='J', ymax=None):
+        if not isinstance(value, Vector3D):
+            raise TypeError("Attribute 'direction' must be of type Vector3D.")
+
+        if value.x != 0 or value.y != 0 or value.z != 1:
```
Yes, we will probably have to add `orientation` property along with `direction` when polarisation is implemented in Raysect. Since we agreed that @CnlPepper will remove unpolarised rays, we will need to update the existing classes rather than create separate ones. However, until then I would leave it as it is.

vsnever commented on cherab/tools/observers/spectroscopic.py:
```diff
@@ -212,11 +656,106 @@ def plot_spectra(self, unit='J', ymax=None, extras=True):
+        return super().pipelines
+
+    @pipelines.setter
+    def pipelines(self, value):
```
If the property is not redefined in the child class, the property of the parent class will be used, and here the parent class allows for multiple pipelines. I will remove this code as we agreed that these observers should support multiple pipelines, both spectral and non-spectral.

vsnever commented:
I added support for multiple pipelines. The pipelines of the following classes  are supported: `PowerPipeline0D`, `RadiancePipeline0D`, `SpectralPowerPipeline0D`, `SpectralRadiancePipeline0D`. Since the property `observed_spectrum` is converted to a method, the code that uses `SpectroscopicSightLine` has to be updated. For example, in cherab/jet/demos/diagnostic/ks5/spectra_76666.py, the code:
```python
for sightline in ks5c.sight_lines:
    observed_radiance.append(sightline.observed_spectrum.total())
```
has to be replaced with:
```python
for sightline in ks5c.sight_lines:
    observed_radiance.append(sightline.observed_spectrum().total())
```
The methods `observed_spectrum()` and `plot_spectra()` now take the index or name of the pipeline as an argument.
However, the pipelines of type `PowerPipeline0D` and `RadiancePipeline0D` cannot be accessed by name until raysect/source#388 is merged.

The pipelines of the same class can be connected to all sight lines in the group with the help of `connect_pipelines()` method. 

This will connect one `SpectralRadiancePipeline0D` named 'Full spectrum' and two `RadiancePipeline0D` named 'Spectral line 1' and 'Spectral line 2' with different filters to all sight lines in the group:
```python
group.connect_pipelines([(SpectralRadiancePipeline0D, 'Full spectrum', None),
                         (RadiancePipeline0D, 'Spectral line 1', filter1),
                         (RadiancePipeline0D, 'Spectral line 2', filter2)])
```
@Mateasek, сould you take another look at the code and tell me if everything is ok now?

Mateasek commented on cherab/tools/observers/spectroscopic.py:
```diff
@@ -60,163 +516,375 @@ def sight_lines(self):
+            if isinstance(pipeline, SpectralPowerPipeline0D):
+                plt.plot(spectrum.wavelengths, spectrum.samples, marker='o', ls='none', label=self.name)
+            elif isinstance(pipeline, PowerPipeline0D):
+                plt.plot(spectrum.wavelengths, spectrum.total(), marker='o', ls='none', label=self.name)
```
To avoid confusion between `SpectralPowerPipeline0D` and `PowerPipeline0D` I would suggest adding only a legend record in case of `PowerPipeline0D`, something like:
plt.plot([], [], marker='none', ls='none', label=self.name + " {} {}".format(value, unit))

Mateasek reviewed (changes requested):
Hi @vsnever ,

thanks for adding the possibility to have multiple Pipelines. I think it will be very useful. I have my concern about the `observed_spectrum` method. I think here also @jacklovell should express his opinion.

Also, this PR is quite large, I'm quite confident that if there are some other problems they escaped me.

vsnever commented on cherab/tools/observers/spectroscopic.py:
```diff
@@ -60,163 +516,375 @@ def sight_lines(self):
+        self.pipelines = pipelines
 
-    def plot_spectra(self, unit='J', ymax=None, extras=True):
+    def observed_spectrum(self, item=0):
```
I agree, `get_pipeline()` is enough. I was just trying to keep all the existing methods.

vsnever commented:
I've dropped the `observed_spectrum()` method to avoid confusion, because as @Mateasek noted the `get_pipeline()` method is enough. 

I've also improved the plotting. The `plot_spectra()` (for a group) and `plot_spectrum()`(for a single LoS) methods now support only spectral pipelines. A new plotting function `plot_total_signal()` can be used with both mono and spectral pipelines. 

Here are the plots generated with `plot_spectra()` and `plot_total_signal()` in the `ks3_group_observation.py` demo from cherab/jet#21:
![KS3_cherab_plot_spectra](https://user-images.githubusercontent.com/22627041/120035190-d91faa80-c006-11eb-986d-d64701799e92.png)
![KS3_cherab_plot_total_signal](https://user-images.githubusercontent.com/22627041/120035202-e046b880-c006-11eb-993c-1b7948f69fbf.png)

vsnever commented:
I want to move this code from `cherab.tools.observer.spectroscopic.py` to `cherab.tools.observer.spectroscopy` (see #300), and create two separate files there, one for `_SpectroscopicObserver0DBase`, `SpectroscopicSightLine` and `SpectroscopicFibreOptic`, and another one for `Observer0DGroup`, `LineOfSightGroup` and `FibreOpticGroup`. The new structure seems clearer to me. @Mateasek, @jacklovell, what do you think?

jacklovell commented:
> I want to move this code from `cherab.tools.observer.spectroscopic.py` to `cherab.tools.observer.spectroscopy` (see #300), and create two separate files there, one for `_SpectroscopicObserver0DBase`, `SpectroscopicSightLine` and `SpectroscopicFibreOptic`, and another one for `Observer0DGroup`, `LineOfSightGroup` and `FibreOpticGroup`. The new structure seems clearer to me. @Mateasek, @jacklovell, what do you think?

It sounds reasonable to move the spectroscopic observers inside `cherab.tools.observer.spectroscopy`. But the `*Group` observers aren't necessarily spectroscopic so should live in their own module `cherab.tools.observer.group`.

vsnever commented:
I've moved the code for each observer from `tools.observers.spectroscopic.py` to a separate file in the `tools.observers.group` or `tools.observers.spectroscopy` modules for a clearer structure. I've also added code examples to the docstrings and updated the `observers.rst`.
